[Site Isolation] Make WKFrameGetWebArchive work under site isolation

szewai opened the pull request:
#### 6bcd6d58a6fabcbe3c874506d9a89c138a80a821
<pre>
[Site Isolation] Make WKFrameGetWebArchive work under site isolation
<a href="https://rdar.apple.com/160665427">rdar://160665427</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298920">https://bugs.webkit.org/show_bug.cgi?id=298920</a>

Reviewed by Ryosuke Niwa.

WKFrameGetWebArchive allows client to get webarchive data for a specified frame. To implement that, this patch makes
UI process iterate all child frames of the specified frame, get their processes, and send messages to all processes to
collect webarchive data for those frames (see `WebPageProxy::getWebArchiveDataWithFrame`).

To test the new implementation, this patch adds a new SPI `_createWebArchiveForFrame`, which uses the same underlying
function as WKFrameGetWebArchive.

To unify the implementation of getting webarchive data with site isolation on and off, this patch also udpates the
implementation of API `createWebArchiveDataWithCompletionHandler` to remove the site isolation enablement check and
make it always invoke `WebPageProxy::getWebArchiveDataWithFrame`. For backward compatibility, this patch retains the
behavior that if the web view has not made any load before the API is invoked, blank page data will be returned.

API Test: SiteIsolation.CreateWebArchiveForFrame

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createWebArchiveDataWithCompletionHandler:]):
(-[WKWebView _createWebArchiveForFrame:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getWebArchiveData):
(WebKit::WebPageProxy::getWebArchiveDataWithFrame):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::getWebArchive):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getWebArchiveOfFrame): Deleted.
(WebKit::WebPageProxy::getWebArchive): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getWebArchivesForFrames):
(WebKit::WebPage::getWebArchiveData):
(WebKit::WebPage::getWebArchives): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveForFrame)):

Canonical link: <a href="https://commits.webkit.org/300098@main">https://commits.webkit.org/300098@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74df85b89a43160e460630de29417d7878539b11

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127795 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73438 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/89f16389-7247-4620-8641-1acdac4df944) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123231 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49631 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92179 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d9cf7b3e-d4ef-4fd5-8cdb-fb7609cc8d89) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124307 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33342 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/108740 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72856 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8ad029d2-af0c-4120-a1c2-dc1566c10b6d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32359 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26878 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71374 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102835 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27053 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130629 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48283 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36710 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100777 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48651 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104938 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100682 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25513 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46093 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24159 "Passed tests") | [💥 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/44977 "An unexpected error occured. Recent messages:Printed configuration") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48141 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53854 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47613 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/50959 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49295 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1670,9 +1670,32 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+    getWebArchiveDataWithFrame(*frame, WTFMove(completionHandler));
+}
+
+void WebPageProxy::getWebArchiveDataWithFrame(WebFrameProxy& frame, CompletionHandler<void(API::Data*)>&& completionHandler)
```
Hm... the function names differ between "for" and "with" and the only difference is its argument type?
Maybe we can call this one `getWebArchiveDataForFrameInternal` instead?

szewai replied:
Yes, `getWebArchiveDataForFrame` is used in WKWebView.mm so it cannot be private -- I will invoke `getWebArchiveDataWithFrame` in WKWebView.mm instead and remove `getWebArchiveDataForFrame`.

szewai committed:
[Site Isolation] Make WKFrameGetWebArchive work under site isolation
rdar://160665427
https://bugs.webkit.org/show_bug.cgi?id=298920

Reviewed by Ryosuke Niwa.

WKFrameGetWebArchive allows client to get webarchive data for a specified frame. To implement that, this patch makes
UI process iterate all child frames of the specified frame, get their processes, and send messages to all processes to
collect webarchive data for those frames (see `WebPageProxy::getWebArchiveDataWithFrame`).

To test the new implementation, this patch adds a new SPI `_createWebArchiveForFrame`, which uses the same underlying
function as WKFrameGetWebArchive.

To unify the implementation of getting webarchive data with site isolation on and off, this patch also udpates the
implementation of API `createWebArchiveDataWithCompletionHandler` to remove the site isolation enablement check and
make it always invoke `WebPageProxy::getWebArchiveDataWithFrame`. For backward compatibility, this patch retains the
behavior that if the web view has not made any load before the API is invoked, blank page data will be returned.

API Test: SiteIsolation.CreateWebArchiveForFrame

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createWebArchiveDataWithCompletionHandler:]):
(-[WKWebView _createWebArchiveForFrame:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getWebArchiveData):
(WebKit::WebPageProxy::getWebArchiveDataWithFrame):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::getWebArchive):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getWebArchiveOfFrame): Deleted.
(WebKit::WebPageProxy::getWebArchive): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getWebArchivesForFrames):
(WebKit::WebPage::getWebArchiveData):
(WebKit::WebPage::getWebArchives): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveForFrame)):

Canonical link: https://commits.webkit.org/300098@main
